src/Microsoft.Azure.WebJobs.Host/Loggers/Logger/Aggregator/FunctionResultAggregator.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Host.Loggers
+{
+    internal class FunctionResultAggregator : IFunctionResultAggregator, IDisposable
+    {
+        private readonly ILogger _logger;
+        private readonly int _batchSize;
+        private readonly TimeSpan _batchTimeout;
+
+        private BufferBlock<FunctionResultLog> _buffer;
+        private BatchBlock<FunctionResultLog> _batcher;
+
+        private Timer _windowTimer;
+        private IDisposable[] _disposables;
+
+        public FunctionResultAggregator(int batchSize, TimeSpan batchTimeout, ILoggerFactory loggerFactory)
+        {
+            _logger = loggerFactory.CreateLogger(LoggingCategories.Aggregator);
+            _batchSize = batchSize;
+            _batchTimeout = batchTimeout;
+            InitializeFlow(_batchSize, _batchTimeout);
+        }
+
+        private void InitializeFlow(int maxBacklog, TimeSpan maxFlushInterval)
+        {
+            _buffer = new BufferBlock<FunctionResultLog>(
+                new ExecutionDataflowBlockOptions()
+                {
+                    BoundedCapacity = maxBacklog
+                });
+
+            _batcher = new BatchBlock<FunctionResultLog>(maxBacklog,
+                new GroupingDataflowBlockOptions()
+                {
+                    BoundedCapacity = maxBacklog,
+                    Greedy = true
+                });
+
+            TransformBlock<IEnumerable<FunctionResultLog>, IEnumerable<FunctionResultAggregate>> aggregator =
+                new TransformBlock<IEnumerable<FunctionResultLog>, IEnumerable<FunctionResultAggregate>>(transform: (e) => Aggregate(e));
+
+            ActionBlock<IEnumerable<FunctionResultAggregate>> publisher = new ActionBlock<IEnumerable<FunctionResultAggregate>>(
+                (e) => Publish(e),
+                new ExecutionDataflowBlockOptions()
+                {
+                    MaxDegreeOfParallelism = 1,
+                    BoundedCapacity = 32
+                });
+
+            _disposables = new IDisposable[]
+            {
+                _buffer.LinkTo(_batcher),
+                _batcher.LinkTo(aggregator),
+                aggregator.LinkTo(publisher)
+            };
+
+            _windowTimer = new Timer(async (o) => await FlushAsync(), null, maxFlushInterval, maxFlushInterval);
+        }
+
+        public Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            _batcher?.TriggerBatch();
```

Comment:
A bit confused on this - does the batch trigger a publish? It seems that FlushAsync will aggregate the current buffer, but I'm not sure when things will publish